File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        print(nums)
        ans = []
        n = len(nums)

        i = 0
        while i<n:
            j = i+1
            k = n-1
            while j<k:
                sum = nums[i]+nums[j]+nums[k]
                if sum==0:
                    temp = [nums[i],nums[j],nums[k]]
                    ans.append(temp)
                    j+=1
                    while j<k and nums[j]==nums[j-1]:
                        j+=1
                    k-=1
                    while j<k and nums[k]==nums[k+1]:
                        k-=1
                elif sum<0:
                    j+=1
                    while j<k and nums[j]==nums[j-1]:
                        j+=1
                else:
                    k-=1
                    while j<k and nums[k]==nums[k+1]:
                        k-=1

            i+=1
            while i<n and nums[i]==nums[i-1]:
                i+=1

        return ans

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Set<List<Integer>> st = new HashSet<>();
        Arrays.sort(nums);
        for(int i=0;i<nums.length;i++){
            int j=i+1;
            int k=nums.length-1;
            while(j<k){
                int sum = nums[i]+nums[j]+nums[k];
                if(sum == 0){
                    List<Integer> temp = new ArrayList<Integer>();
                    temp.add(nums[i]);
                    temp.add(nums[j]);
                    temp.add(nums[k]);
                    st.add(temp);
                }

                if(sum<0){
                    while(j<k && nums[j]==nums[j+1])j++;
                    j++;
                }
                else{
                    while(j<k && nums[k]==nums[k-1])k--;
                    k--;
                }
            }

            while(i+1<nums.length && nums[i]==nums[i+1])i++;
        }

        return new ArrayList<>(st);
    }
}